requests-for-comments/2022-11-02-data-exploration.md

```diff
@@ -0,0 +1,600 @@
+# Request for comments: Data Exploration - Marius Andra
+
+_Whenever people ask me to be the life of the party, I go into a long rant about the importance of getting your data structures right.
+
+It's the single most impactful advice you can give junior developers, and it's what we're going to do here._
+
+## Why?
+
+We want to build a "data exploration" view, which lets users take any data for any insight, and explore it further.
+
+Users have been asking for this for a while. Some example pain points:
+
+- You can't drill down on insight person modals (e.g. filter the list further, break down by cohorts, etc).
+- Our tables are basic. E.g. The events table supports some customisation, but not enough. What about aggregations? Tables below insights contain different data in insight vs table mode. Persons tables are very basic. Math and formulas?
+- There are chart components that can't be added to dashboards (e.g. retention line chart)
+- Various code reasons (explained below) of why it's a bad idea to keep appending stuff to `FilterType`
+- Multiple endpoints accept various types of filters with various capabilities (`/insights/funnel/correlations` vs `/insights/trend/` vs `/events` vs `/persons/trends`), hard to know what is supported where.
+
+## What?
+
+This is a proposal to change the way we ask for data from the PostHog API. This will unblock the new "data exploration" view, and also:
```

Comment:
I'm curious why you think changing the way we ask for data is the core component of data exploration here?

It seems to suggest that what's hard about this project is "asking for what you want" & how to express that. While useful to have a more uniform query type, I'm not convinced that's where all the complexity lies here.

Rather, it _seems_ to me to be a nice to have (at this stage of the project). I think the complexity lies in trying to combine diverse filters together in a query.

I might be completely off the mark here (or in the end the solution might just become what you propose), but I think this MVP should focus more on validating / figuring out how diverse backend filters can interact together.

Reply:
Thanks for asking! This is exactly what the RFC is for: to discuss if we're tackling the right problems, and in the right order. However this is also still "brainstorming mode", where everything's possible and there are no really stupid answers... so treat this more as an "improv comedy" (more "yes and", less "yes but") 😅.

> I'm curious why you think changing the way we ask for data is the core component of data exploration here?

Well that's because if you start with data, then everything else falls into place easier. 😆 

If in this RFC, as the first step, I would have shown just a lot of screenshots of "[universal search](https://github.com/PostHog/posthog/issues/7963)", the mockups from our offsite design sprint, or went into the specifics on how a `breakdown on paths` differs from `breakdown on funnels` at the clickhouse SQL level, we would have gotten lost in detail. I'd like to get the big abstractions out of the way first, *even if we implement them last*. This is a markdown file asking for exactly these comments after all, not 8 man-hours of code :).

The big picture if we don't want to accrue technical debt before the next refactor in 2 years: **the data exploration view should be declarative, composable and embeddable.**

Now let's figure out how all our existing filters clash with this abstraction, and how we can combine existing filters together in a query. I don't have all the answers, but I'll start exploring this next, along with some frontend ideas.

Reply:
Food for thought:

If we suspend our understanding of the current system and focus on answering the question: how do we enable any sort of behavioral analysis question, I would draw a line between trends and [funnels, retention, lifecycle, paths, stickiness]. Trends, apart from `unique users`, counts events and every other insight counts users [persons, groups]. You're left with a bunch of variations of "funnels" or event sequences. 

Funnels = counting users who have done X events in order
Retention = counting users who have done 2 events in order over recurring period
Stickiness = counting users who have done 2 events in order where the second event is unbounded and you count
Paths = counting users who have coinciding sequence of events
Lifecycle = counting users by how they performed an event at specific times over the entire course of being tracked

Cohort expressions:
<img width="281" alt="Screen Shot 2022-11-03 at 7 53 22 PM" src="https://user-images.githubusercontent.com/13127476/199856481-809dff73-01ad-4369-b092-4b2cdcc4b09a.png">

<img width="277" alt="Screen Shot 2022-11-03 at 7 53 42 PM" src="https://user-images.githubusercontent.com/13127476/199856514-2d99addd-7e3e-48c1-b4fc-48277ea8ce6e.png">

So, what's been done so far is trying to focus on making the highest value event sequence queries accessible and filterable. If you ask how to make this limitlessly expressible, you would need an input that gives you the ability to specify any number of events in sequence, the period which the events occur, the relation of time between any of the events in sequence, the count of each event in the sequence, and probably some more parameters here.

**The interface here becomes another challenge.

I'm skeptical how reasonable it is to chase a system like this atm. Mainly adding here for context on the far reaches of expressiveness in behavioral queries and it adds more depth to how questions can be asked. The initial idea proposed suggests that "types" are trends, funnels, retention, paths, etc but categorically it's events aggregated in specific ways and person sequences aggregated in specific ways. For example, how would we extend the types to include "users who did an event A stopped for X amount of time and restarted doing by event B"? We'd have to add a different type for every configuration of event sequences that is being asked for

I'll +1 what @neilkakkar mentioned above that the first step is really `I think this MVP should focus more on validating / figuring out how diverse backend filters can interact together, & whether all of them should interact together.` How does a query `property breakdown in a [path step, retention, lifecycle]` even look, does it work, and will it be performant?